fecgov/openFEC
# [WIP] Add /communication_costs/ endpoint with CommunicationCostPageSchema pagination schema

## Summary (required)

- Resolves #4203 
- api endpoints that support pagination through page, some. endpoints make the page key explicitly available (e.g., /communication_costs/totals/by_candidate/) while others do not (e.g., /communication_costs/).

For uniformity and to support generalizing programmatic access to our API, we should make pagination keys when possible (a case where this would not be currently possible, for example, would be the schedule_a endpoints where additional keys are needed to support pagination.

### Background
There are 4 endpoints under the communication cost tag:

/committee/{committee_id}/communication_costs/by_candidate/, /communication-costs/, /communication_costs/by_candidate/, ​/communication_costs​/totals​/by_candidate​/. Only /communication-costs/ does not have an explicit page key in pagination. Will investigate what is different about the paging for this endpoint and any rationale.

The /communication-costs/ endpoint currently uses the SeekPageSchema. These are the other schemas using SeekPageSchema (from marshmallow-pagination):

./schemas.py:ScheduleAPageSchema = make_page_schema(ScheduleASchema, page_type=paging_schemas.SeekPageSchema)
./schemas.py:ScheduleBPageSchema = make_page_schema(ScheduleBSchema, page_type=paging_schemas.SeekPageSchema)
./schemas.py:ScheduleH4PageSchema = make_page_schema(ScheduleH4Schema, page_type=paging_schemas.SeekPageSchema)
./schemas.py:ScheduleEPageSchema = make_page_schema(ScheduleESchema, page_type=paging_schemas.SeekPageSchema)
./schemas.py:ElectioneeringPageSchema = make_page_schema(ElectioneeringSchema, page_type=paging_schemas.SeekPageSchema)

The issue here is that a user reported inconsistent pagination between communications cost endpoints (there are 4). Three endpoints use the OffsetPageSchema from marshmallow-pagination while the base endpoint /communication-costs/ uses the SeekPageSchema. The difference is that the OffsetPage uses page number for pagination, while the SeekPage uses the keyset pagination found in Schedule A, for example.

When no argument is supplied to make_page_schema the default is OffsetPageSchema. The majority of our endpoints use this pagination schema. I found that the SeekPageSchema appears to be used in conjunction with our large dataset tables, e.g., Schedule A. These page schemas were implemented prior to my supporting the API, so I'm copying you in case there is any institutional memory for the motivation for choosing the particular paging schema (SeekPageSchema) for the /communication-cost/ endpoint while using the more common OffsetPageSchema for the others.

If there is a specific reason we can probably close this issue and note it in case it comes up again. If there is not, I'd suggest we bring it in line with the other endpoints and notify API users of a potential breaking change in the pagination. If this is not acceptable, perhaps we can create another endpoint that duplicates /communication-cost/ with the new endpoint using the more common paging schema.

### How to test
- checkout branch
- run local api
- compare page schema to see if acceptable

### Additional considerations
- decide if this change is acceptable to implement as is
  - yes: notify api users as appropriate through email/documentation
  - no: decide if issue should be closed with original reporting individual notified or if another endpoint should be created to satisfy consistency.
